src/autopas/containers/hierarchicalGrids/HierarchicalGrids.h

```diff
@@ -0,0 +1,644 @@
+
+      }
+
+  /**
+   * @copydoc autopas::ParticleContainerInterface::getRegionIterator()
+   */
+  [[nodiscard]] ContainerIterator<ParticleType, true, true> getRegionIterator(
+      const std::array<double, 3> &lowerCorner, const std::array<double, 3> &higherCorner, IteratorBehavior behavior,
+      typename ContainerIterator<ParticleType, true, true>::ParticleVecType *additionalVectors = nullptr) override {
+
+        return _carrierLevel.getRegionIterator(lowerCorner, higherCorner, behavior, additionalVectors);
+
+      }
+
+  /**
+   * @copydoc autopas::ParticleContainerInterface::getRegionIterator()
+   * @note const version
+   */
+  [[nodiscard]] ContainerIterator<ParticleType, false, true> getRegionIterator(
+      const std::array<double, 3> &lowerCorner, const std::array<double, 3> &higherCorner, IteratorBehavior behavior,
+      typename ContainerIterator<ParticleType, false, true>::ParticleVecType *additionalVectors = nullptr) const override {
+
+        return _carrierLevel.getRegionIterator(lowerCorner, higherCorner, behavior, additionalVectors);
+
+      }
+
+  template <typename Lambda>
+  void forEachInRegion(Lambda forEachLambda, const std::array<double, 3> &lowerCorner,
+                       const std::array<double, 3> &higherCorner, IteratorBehavior behavior) {
+    
+    _carrierLevel.forEachInRegion(forEachLambda, lowerCorner, higherCorner, behavior);
+
+    }
+
+  template <typename Lambda, typename A>
+  void reduceInRegion(Lambda reduceLambda, A &result, const std::array<double, 3> &lowerCorner,
+                      const std::array<double, 3> &higherCorner, IteratorBehavior behavior) {
+    
+    _carrierLevel.reduceInRegion(reduceLambda, result, lowerCorner, higherCorner, behavior);
+
+    }
+
+
+  void distributeParticlesToHGLevels() {
+
+      AUTOPAS_OPENMP(parallel)
+      for (auto iterParticle = _carrierLevel.begin(); iterParticle.isValid(); ++iterParticle) {
+        // get level the particle should be inserted to
+        unsigned int targetLevel = getHierarchyLevelOfParticle(*iterParticle);
+
+        // add particles to their respective levels
+        addParticleToGridLevel(*iterParticle, targetLevel);
+      }
+      
+      _carrierLevel.deleteAllParticles();
+
+  }
+
+  void distributeParticlesToCarrierLevel() {
+
+    //iterate over all HG levels
+    for (unsigned int iterLevels = 0; iterLevels < _numberOfLevels; iterLevels++) {
+      
+      //iterate over all particles in a given level
+      for(auto iterParticle = _hierarchyLevels[iterLevels].begin(); iterParticle.isValid(); ++iterParticle) {
+
+        //add the particle to the carrier level
+        _carrierLevel.addParticleImpl(*iterParticle);
+
+      }
+      _hierarchyLevels[iterLevels].deleteAllParticles();
+
+    }
+
+  }
+
+  void iterateAllHGLevels(TraversalInterface *traversal) {
+
+    //AUTOPAS_OPENMP(parallel firstprivate (traversal))
+    for (auto & hgLevel : _hierarchyLevels) {
+
+      hgLevel.iteratePairwise(traversal);
```

Comment:
This could be an interesting optimization: Reduce the number of threads per level and parallelize as much as possible over the levels.

But I fear this might mean some major redesign all over autopas? :shrug:

Reply:
Added comment that this is a possibility